# Bologna Transit Analysis - Step 02: Analisi Macroscopica e Proprieta' Small-World

Questo notebook valuta le proprietà macroscopiche globali della rete. Confrontiamo la rete bus con quella comprensiva di tram.
Verifichiamo:
- **Average Shortest Path Length ($L$)**: Numero tipico di passaggi tra stazioni.
- **Clustering Coefficient ($C$)**: Presenza di gruppi o loop locali.
- **Efficienza Globale ($E_{glob}$)**: Capacità di trasporto in parallelo su tutta la rete.
- **Coefficiente $\sigma$ e $\omega$**: Per stabilire se la rete mostra un comportamento Small-World rispetto a modelli nulli casuali (Erdős-Rényi) e regolari (Lattice).

In [1]:
import sys
sys.path.append('..')
from src.data_loader_bologna import load_bologna_network
from src.analyzer_bologna import compute_small_worldness

## 1. Calcolo dei Coefficienti Globali

In [2]:
G_bus, _, _, _, _ = load_bologna_network(include_tram=False)
G_tram, _, _, _, _ = load_bologna_network(include_tram=True)

sw_bus = compute_small_worldness(G_bus)
sw_tram = compute_small_worldness(G_tram)

print(f"{'Metrica':<30}{'Rete Bus':<15}{'Rete Bus+Tram':<15}")
print("-" * 60)
print(f"{'Avg Path Length L':<30}{sw_bus['L']:.4f}{'':<8}{sw_tram['L']:.4f}")
print(f"{'Clustering Coeff C':<30}{sw_bus['C']:.4f}{'':<8}{sw_tram['C']:.4f}")
print(f"{'Global Efficiency E_glob':<30}{sw_bus['E_glob']:.4f}{'':<8}{sw_tram['E_glob']:.4f}")
print(f"{'Local Efficiency E_loc':<30}{sw_bus['E_loc']:.4f}{'':<8}{sw_tram['E_loc']:.4f}")
print(f"{'Small-Worldness Sigma':<30}{sw_bus['sigma']:.4f}{'':<8}{sw_tram['sigma']:.4f}")
print(f"{'Small-Worldness Omega':<30}{sw_bus['omega']:.4f}{'':<8}{sw_tram['omega']:.4f}")

Metrica                       Rete Bus       Rete Bus+Tram  
------------------------------------------------------------
Avg Path Length L             3.7065        3.8832
Clustering Coeff C            0.1378        0.1389
Global Efficiency E_glob      0.2335        0.2829
Local Efficiency E_loc        0.1570        0.1700
Small-Worldness Sigma         1.5267        1.3902
Small-Worldness Omega         0.0000        0.0000


## 2. Interpretazione dei Risultati

- **Sigma ($\sigma > 1$)**: Conferma le proprietà Small-World di Bologna. Entrambe le configurazioni mostrano un clustering superiore al modello nullo casuale, mantenendo cammini brevi.
- **Efficienza Globale**: L'inserimento del Tram fa balzare l'efficienza globale ($E_{glob}$) da `0.2335` a `0.2829`, dimostrando che i collegamenti rapidi su sede protetta aumentano l'integrazione di rete.